### ISSUE TITLE
"Go To References" for `PartialOrd` functions should find comparison operators

### ISSUE THREAD

**Issue by JarredAllen** [rust-lang/rust-analyzer#18540]
If I have code like this:

```rust
pub struct Foo { .. }

impl PartialOrd for Foo { .. }

// also `Eq` and `PartialEq` are implemented consistently with `PartialOrd`

fn compare_foos(foo1: Foo, foo2: Foo) -> bool {
    foo1 <= foo2
}
```

Running "go to definition" on the `<=` will go to the definition of the `le` method in the `impl PartialOrd for Foo` (including to the default implementation in the `PartialOrd` trait if not explicitly overridden in the `impl` block), but running "go to references" can't go back the other way.

It'd be great if I could "go to references" on the `le` function to find all places where `<=` is used to compare 2 `Foo`s. It'd be even greater if, for a type that only implements the `partial_cmp` function and leaves the default implementations alone, "go to references" could find all the `<`/`<=`/`>`/`>=` comparison operators (not sure if this second part is feasible, since it does technically go through a different method).

**Comment by ChayimFriedman2** [comment]
Ah yes I noticed that a while ago. Do note however that this is going to be pretty much the slowest thing we ever provide (find all references is based on textual search then refinement using semantics info - because calculating semantic info is expensive. Searching for operators is going to find *a lot* of matches).

**Comment by 0xorial** [comment]
I have a case which seems very similar:

```
use bevy::prelude::Resource;

#[derive(Resource)]
pub struct UiState {
}

impl Default for UiState {
    fn default() -> Self {
        UiState {}
    }
}
```

and then somewhere else 

```
.init_resource::<UiState>()
```

If I understand correctly it is known in compile time that this specific Default implementation is going to be used in that specific location. But trying to find references on "Default" just gives all thousands of references to Default trait, not to this implementation...

**Comment by Veykril** [comment]
That is not relevant to this issue, please open a new one with that and be more descriptive with what you are attempting and what you would expect as its not clear to me from your comment.

**Comment by svix-jplatte** [comment]
I was looking for `<chrono::DateTime<Utc> as Add<chrono::TimeDelta>>::add` references today (after seeing a panic from one such addition that had no backtrace for reasons I do not yet understand). I'm pretty sure we don't have many of these in the codebase. I understand "find all references" for this will take a long time, but right now I have no tools to find all of these at all, short of manually going through all `+` symbols in the codebase which is obviously infeasible.

**Comment by ChayimFriedman2** [comment]
This is going to be *very* slow (my lowest estimate is in order of minutes). If people still find value in that, I'm not opposed to it, as there isn't really a maintenance burden.

**Comment by svix-jplatte** [comment]
It would be awkward if it was more than about a minute, but better than nothing IMO. Maybe it could be brought down by only looking inside the workspace (for find-all-references in general)? Right now, I'm getting results from find-all-references that are in upstream crates, and I think sysroot crates too.

**Comment by ChayimFriedman2** [comment]
Limiting reference search to only some files is an interesting idea in general. The UI is a problem though.

(And FWIW I meant *several* minutes as my *lowest* estimate (for a medium-large codebase). It could take even more).

**Comment by svix-jplatte** [comment]
> Limiting reference search to only some files is an interesting idea in general. The UI is a problem though.

Yeah, I know. Seems like something I'd want to change rarely enough that a global setting would work, but maybe that's too awkward? Happy to create a new ticket about this if useful. 

> (And FWIW I meant *several* minutes as my *lowest* estimate (for a medium-large codebase). It could take even more).

Yeah I got that. I'd still want to see it. If you can provide some rough instructions I might find some time to create a PR :)

**Comment by ChayimFriedman2** [comment]
Our search infra is in: https://github.com/rust-lang/rust-analyzer/blob/2a16118416ca66c19686ea63869039c056cb1ea2/crates/ide-db/src/search.rs. You will need to detect when searching for an operator method (possible by looking at the function's trait and checking its `#[lang]` item, you can draw inspiration from `hir::Trait::lang()` that does the opposite - takes a lang item and returns the trait), implement searching of operators textually, then you can verify the results by calling `OperatorClass` classification methods, similar to how existing verification does it for `NameClass` and `NameRefClass`.
